src/backends/perf_pt/perf_pt.c

```diff
@@ -105,81 +105,66 @@ struct tracer_thread_args {
 };
 
 
 // Private prototypes.
-static bool write_buf_to_disk(int, void *, __u64);
-static bool read_circular_buf(void *, __u64, __u64, __u64 *, int);
-static bool poll_loop(int, int, int, struct perf_event_mmap_page *, void *);
+static bool read_aux(void *, __u64, __u64, __u64 *, void **, __u64, __u64 *);
+static bool poll_loop(int, int, struct perf_event_mmap_page *, void *, void **,
+                      __u64, __u64 *);
 static void *tracer_thread(void *);
 static int open_perf(pid_t target_tid);
 
 // Exposed Prototypes.
 struct tracer_ctx *perf_pt_start_tracer(struct tracer_conf *);
-int perf_pt_stop_tracer(struct tracer_ctx *tr_ctx);
-
+int perf_pt_stop_tracer(struct tracer_ctx *tr_ctx, uint8_t **, __u64 *);
+void *perf_pt_trace_buf(struct tracer_ctx *tr_ctx, uint32_t *len);
 
 /*
- * Write part of a buffer to a file descriptor.
+ * Read data out of the AUX buffer.
  *
- * The implementation is a little convoluted due to friction between ssize_t
- * and __u64.
- *
- * Returns true on success or false otherwise.
- */
-static bool
-write_buf_to_disk(int fd, void *buf, __u64 size) {
-    char *buf_c = (char *) buf;
-    size_t block_size = SSIZE_MAX;
-
-    while (size > 0) {
-        if (block_size > size) {
-            block_size = size;
-        }
-        ssize_t res = write(fd, buf, block_size);
-        if (res == -1) {
-            if (errno == EINTR) {
-                // Write interrupted before anything written.
-                continue;
-            }
-            return false;
-        }
-        size -= res;
-        buf_c += res;
-    }
-    return true;
-}
-
-/*
- * Read data out of a circular buffer.
+ * Reads `size` bytes from `aux_buf` into `trace_buf`, updating `*trace_len`.
  *
  * Returns true on success and false otherwise.
  */
 static bool
-read_circular_buf(void *buf, __u64 size, __u64 head_monotonic, __u64 *tail_p, int out_fd) {
+read_aux(void *aux_buf, __u64 size, __u64 head_monotonic, __u64 *tail_p,
+         void **trace_buf, __u64 trace_bufsize, __u64 *trace_len) {
     __u64 head = head_monotonic % size; // Head must be manually wrapped.
     __u64 tail = *tail_p;
 
+    // First check we won't overflow the (destination) trace buffer.
+    __u64 new_data_size;
     if (tail <= head) {
         // No wrap-around.
-        DEBUG("read with no wrap");
-        if (!write_buf_to_disk(out_fd, buf + tail, head - tail)) {
-            return false;
-        }
+        new_data_size = head - tail;
     } else {
         // Wrap-around.
+        new_data_size = (size - tail) + head;
+    }
+
+    // For now we simply crash out if we exhaust the buffer.
+    DEBUG("current trace length: %lld", *trace_len);
+    DEBUG("new_data_size: %lld", new_data_size);
+    DEBUG("trace_bufsize: %lld", trace_bufsize);
+    assert(*trace_len + new_data_size <= trace_bufsize);
+
+    // Finally copy the AUX buffer into the trace buffer.
+    if (tail <= head) {
```

Comment:
Can `tail == head`?

Reply:
yes, but it's ambiguous. Either the buffer is empty or full. This is related to #4.

Reply:
Until we implement that, it's assumed to be empty.

Reply:
You've lost me.

Reply:
The aux buffer is a ringer buffer.

It starts with `head == tail`, and as data goes in, `head` zips along the buffer. If you don't take data out quick enough, it eventually wraps around. PT provides a way to know if this happens, but we have not implemented it yet.

Reply:
[if it happens, then we lost packets]